### ISSUE TITLE
Root cause why some presubmit jobs are not finishing

### ISSUE THREAD

**Issue by hklai** [istio/test-infra#957]
See https://k8s-gubernator.appspot.com/pr/istio/istio/8347 and https://k8s-gubernator.appspot.com/pr/istio/istio/8292

There are a few presubmit jobs that were never finished, and newer runs were started. 

According to @BenTheElder, it could be due to rebasing/force push, or the test pod gets rescheduled. 
If the nodes are not stable then kubernetes will re-schedule the pod, and it will collect a new build ID. It had happened in k8s when a broken workload caused the nodes to run out of PIDs. 

"if these were not prow jobs, and just some random kubernetes pod, then if the node becomes unhealthy kubernetes restarts or reschedules the pod since prowjobs are usually just kubernetes pods managed by a prow controller, this can happen to them too. if this does happen to them, they will re-vend the build ID on boot IIRC. See https://github.com/kubernetes/test-infra/issues/5700."

"if it's happening something is wrong with the cluster possibly
we haven't had any real issues with this for some time now, and we were able to root cause those to a specific workload that was misbehaving badly and causing issues with the nodes
https://github.com/kubernetes/test-infra/issues/5877
(that was the root cause for most of the problems we saw)"

Since there was one commit in #8347 and there is no force push during today, and it may be useful to figure out if pod rescheduling is our problem and whether we can fix it.

Seb, could you check the prow log?

**Comment by BenTheElder** [comment]
> Since there was one commit in #8347 and there is no force push during today

Not even by individual PR authors to their own PRs?

**Comment by ldemailly** [comment]
nope, no change whatsoever to https://github.com/istio/istio/pull/8347 past initial submission yet it's done presubmit 5 times so far and taken > 3h

**Comment by ldemailly** [comment]
also 3x the green one when only 1 PR got merged in master in the window is a separate but also serious issue/bug

**Comment by BenTheElder** [comment]
This is likely a problem with the Kubernetes cluster then, jobs are only automatically aborted and rescheduled when:

- you push new code to the PR while a job is running, so Prow starts a new one instead and aborts the old one
- Kubernetes has to re-schedule the test pod (possibly evicted)

The latter is not specific to prow and not really debuggable without looking at the cluster's logs.

**Comment by ldemailly** [comment]
@hklai want to file a separate issue for the triplicate green builds:
![screen shot 2018-08-29 at 4 28 04 pm](https://user-images.githubusercontent.com/3664595/44821094-9f48bd80-aba8-11e8-92b9-4b5dd50e3188.png)

**Comment by ldemailly** [comment]
it took over 12 hours, 14 presubmit, ~60+ other builds to finally get in

<img width="1113" alt="screen shot 2018-08-30 at 10 37 27 am" src="https://user-images.githubusercontent.com/3664595/44868855-e5e9f680-ac40-11e8-9341-9d87157eb934.png">

**Comment by sebastienvas** [comment]
Apparently the looping issue is actually coming from Tide. We have after_success jobs that are always_run: false (because they get triggered). As soon as a PR is ready to be merged, it triggers the tests, and the PR disappear from Tide until the always_run: false are passing in which case it comes back to tide, which re-trigger a new set of tests.

There might something I don t fully understand but I am just reporting what I have been seeing, while talking to @cjwagner. Tide does not properly supportes run after success job.

I updated all the after_success to be always_run: true to see if that makes any difference, but PR still does not show up in tide.

**Comment by cjwagner** [comment]
> I updated all the after_success to be always_run: true to see if that makes any difference, but PR still does not show up in tide.

Yeah, this won't fix things. The core issue is that Tide knows the run_after_success jobs are required, but doesn't know how to trigger them. I had a PR to add support for this to Tide, but it doesn't sound like something we want to go through with:  https://github.com/kubernetes/test-infra/pull/8868

**Comment by sebastienvas** [comment]
How do you guys deal with building something that is required by multiple e2e ? Do you build it on every job ?

**Comment by BenTheElder** [comment]
we are generally continuously testing HEAD at some rate for each suite, the
tests download the latest uploaded build and log which build commit they
used.
this may add some small latency for the build to complete, but it's simple
and gives us valuable data on test flakiness, and tests running on a cron /
interval don't give us load spikes.

**Comment by sebastienvas** [comment]
Yeah, but I don t see how this applies for PRs ?

**Comment by BenTheElder** [comment]
PRs use [cached builds](https://github.com/kubernetes/test-infra/tree/master/greenhouse) near exclusively so we're not concerned about building on more than one job.

**Comment by sebastienvas** [comment]
if all the builds run in parallel, the cache only improves incremental builds and re-runs, which helps for sure.  But in istio case, if we want to get rid of after_success, we need to run the docker builds in every e2e, which would waste a lot of CPUs. Local golang and docker cache (on ssd) might help a little bit, but those are not in place yet.

I am going to try Cole's Tide fix to see if it improves. @fejta, heads up, I might need your help get that PR in if this ends up working for us. More to come once I have an update.

Do we have an ETA on the pipeline CRDs ?

**Comment by BenTheElder** [comment]
> if all the builds run in parallel, the cache only improves incremental builds and re-runs, which helps for sure

Not true. The Bazel cache is concurrent and globally shared on every action tree.

> Local golang and docker cache (on ssd) might help a little bit, but those are not in place yet.

Yeah but concurrency and linearization are an issue with straightforward disk caching. I'd be interested in what you come up with or possibly discussing rough ideas for how Prow could do this better. I've been mulling this problem for other builds.

> Do we have an ETA on the pipeline CRDs ?

The build CRD? There's rough support now.

**Comment by sebastienvas** [comment]
So its seems that Cole's change actually improve our workflow. I would like to have it run for a day or more but What do you guys advise to do about the PR ?

**Comment by fejta** [comment]
This seems super stale.

If the job isn't finished its probably because: the job doesn't define a timeout and/or the pod failed unexpectedly for some reason (eviction)
